Feature/api key endpoint

MikeNeilson opened the pull request:
Decided to move the User API key generation to the API.

This is a user operation after authentication so I think it's reasonable to present here. Most systems, like say BitBucket, work that way so we aren't way out in left field here.

MikeNeilson committed:
bump version.

MikeNeilson committed:
Pull in Location Level unit fix.

MikeNeilson committed:
Bump version.

adamkorynta committed:
only retrieve rating metadata for non alias locations

adamkorynta committed:
store xml without converting to RatingSet data object first

this will allow us to store single rating curves without specs or templates defined.

MikeNeilson committed:
add required variables for POST authorization.

MikeNeilson committed:
initial work on Api Key endpoint.

MikeNeilson committed:
Can create API key.

MikeNeilson committed:
Work in progress on key list.

MikeNeilson committed:
can list keys.

MikeNeilson committed:
Better logging setup.

MikeNeilson committed:
Initial delete implementation.

MikeNeilson committed:
Better messaging.

MikeNeilson committed:
Reset version back to snapshot.

MikeNeilson commented:
Not sure why the rating spec changes got in here; however they are required and allow those tests to actually run so I'm going to just go with it.

MikeNeilson committed:
Additional test and additional log if failed print.

adamkorynta commented:
This would allow us to create an API Key GUI in CWMS clients if we wanted to, as opposed to prompting the user to open a webpage in the browser.

MikeNeilson commented:
Yep, which could then also be used for longer session. Wouldn't solve the "I want my model to have run in the morning" issue since that's the IT department restarting the computer. But it does make some other things easier.

DanielTOsborne commented on cwms-data-api/src/main/java/cwms/cda/api/RatingController.java:
```diff
@@ -134,24 +139,18 @@ private ContentType getContentType(Context ctx) {
-    }
-
-    public RatingSet deserializeRatingSet(String body, String contentType) throws IOException,
+    String deserializeRatingSet(String body, String contentType) throws IOException,
```
Shouldn't this be renamed to "serializeRatingSet"?
Seems like the operation was inverted.

MikeNeilson replied:
I'm not going to tweak the name. I think some of the confusion stems from how the libraries we use work.

Plus this just needs to get into `develop`, I was a little surprised to see it in this branch but it is needed for the Rating tests to work.

DanielTOsborne commented on cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java:
```diff
@@ -52,7 +61,11 @@ public class AuthDao extends Dao<DataApiPrincipal>{
 
-
+    public static final String CREATE_API_KEY = "insert into cwms_20.at_api_keys(userid,key_name,apikey,created,expires) values(UPPER(?),?,?,?,?)";
+    public static final String REMOVE_API_KEY = "delete from cwms_20.at_api_keys where UPPER(userid) = UPPER(?) and key_name = ?";
```
If AT_API_KEYS table has an index, does it have a function-based index on UPPER(userid)?

MikeNeilson replied:
Userid is ALWAYS already uppercase. So i might not need the function here but it seemed clearer... and it's possible that will change in the future.

DanielTOsborne commented on cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java:
```diff
@@ -250,4 +263,116 @@ private static String getFailMessage(@NotNull Context ctx,
+                throw new CwmsAuthException(ONLY_OWN_KEY_MESSAGE, HttpCode.UNAUTHORIZED.getStatus());
+            }
+            SecureRandom randomSource = SecureRandom.getInstanceStrong();
+            String key = randomSource.ints(48,122) // allow a-zA-Z0-9
```
Personally, I'd probably use something like
`randomSource.ints((char)'0',(char)'z')`
to make it clearer to the reader what the range, and exclusions are.

DanielTOsborne commented on cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java:
```diff
@@ -250,4 +263,116 @@ private static String getFailMessage(@NotNull Context ctx,
+                    createKey.setString(1,newKey.getUserId());
+                    createKey.setString(2,newKey.getKeyName());
+                    createKey.setString(3,newKey.getApiKey());
+                    createKey.setDate(4,new Date(newKey.getCreated().toInstant().toEpochMilli()));
```
The date will be interpreted in the JVM's timezone. Pass a Calendar object for the 3rd argument to enforce UTC interpretation.
For all set/getDate calls.
https://docs.oracle.com/javase/8/docs/api/java/sql/PreparedStatement.html#setDate-int-java.sql.Date-java.util.Calendar-

MikeNeilson replied:
Thanks. that likely would've bit me a some point.

DanielTOsborne commented on cwms-data-api/src/test/java/cwms/cda/api/auth/ApiKeyControllerTestIT.java:
Is there a test for using an already expired key?
I couldn't see any actual date range validation code.

Or, a non-existent key.

MikeNeilson replied:
Non-existent yes, though I think it's buried in there. I don't think I checked for an expired key though. I'll add that.

MikeNeilson committed:
Review updates

Enforce time zone correctly for setDate.
Alter randomSource.ints for clarity.
Additional tests for expired and non-existent api keys.

MikeNeilson commented:
@DanielTOsborne changes made.
